python/cudf/cudf/core/column/column.py

```diff
@@ -2032,117 +2037,74 @@ def as_column(
             return ColumnBase.from_scalar(arbitrary, length)
 
     elif hasattr(arbitrary, "__array_interface__"):
-        # CUDF assumes values are always contiguous
         desc = arbitrary.__array_interface__
-        shape = desc["shape"]
-        arb_dtype = np.dtype(desc["typestr"])
+        check_invalid_array(desc["shape"], np.dtype(desc["typestr"]))
+
         # CUDF assumes values are always contiguous
-        if len(shape) > 1:
-            raise ValueError("Data must be 1-dimensional")
+        arbitrary = np.asarray(arbitrary, order="C")
 
-        arbitrary = np.asarray(arbitrary)
+        if arbitrary.ndim == 0:
+            # TODO: Or treat as scalar?
+            arbitrary = arbitrary[np.newaxis]
 
-        # Handle case that `arbitrary` elements are cupy arrays
-        if (
-            shape
-            and shape[0]
-            and hasattr(arbitrary[0], "__cuda_array_interface__")
-        ):
+        if arbitrary.dtype.kind in "OSU":
+            # Handle case that `arbitrary` elements are cupy arrays
+            if len(arbitrary) > 0 and hasattr(
+                arbitrary[0], "__cuda_array_interface__"
+            ):
+                return as_column(
+                    cupy.asarray(arbitrary, dtype=arbitrary[0].dtype),
+                    nan_as_null=nan_as_null,
+                    dtype=dtype,
+                    length=length,
+                )
+            if pd.isna(arbitrary).any():
+                arbitrary = pa.array(arbitrary)
+            else:
+                arbitrary = pd.Series(arbitrary)
```

Comment:
What happens if we just always use pyarrow here? Do we lose something from non-nullable data that pandas captures better?

Reply:
There are cases where pandas will infer `dtype=object` into a non-object type that would help match pandas behavior e.g. 

```python
In [1]: import cudf, pandas as pd, numpy as np

# e.g a `to_numpy()` round-trip
In [2]: pd.Series(np.array([pd.Timestamp(2020, 1, 1)], dtype=object))
Out[2]: 
0   2020-01-01
dtype: datetime64[ns]
```

I'll add a comment about that here